toplevel/identity/entity.go

```diff
@@ -0,0 +1,642 @@
+package identity
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/app-sre/vault-manager/pkg/utils"
+	"github.com/app-sre/vault-manager/pkg/vault"
+	"github.com/app-sre/vault-manager/toplevel"
+	"github.com/hashicorp/go-version"
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+)
+
+type config struct{}
+
+var _ toplevel.Configuration = config{}
+
+type user struct {
+	Name        string `yaml:"name"`
+	OrgUsername string `yaml:"org_username"`
+	Roles       []role `yaml:"roles"`
+}
+
+type role struct {
+	Name        string           `yaml:"name"`
+	Permissions []oidcPermission `yaml:"oidc_permissions"`
+}
+
+type oidcPermission struct {
+	Name        string        `yaml:"name"`
+	Description string        `yaml:"description"`
+	Service     string        `yaml:"service"`
+	Policies    []vaultPolicy `yaml:"vault_policies"`
+}
+
+type vaultPolicy struct {
+	Name string `yaml:"name"`
+}
+
+type entity struct {
+	Name     string
+	Id       string
+	Type     string
+	Metadata interface{}
+	Policies []string
+	Aliases  []entityAlias
+}
+
+type entityAlias struct {
+	Name           string
+	Id             string
+	Type           string
+	AuthType       string
+	AccessorId     string
+	CustomMetadata interface{}
+}
+
+var _ vault.Item = entity{}
+
+var _ vault.Item = entityAlias{}
+
+func (e entity) Key() string {
+	return e.Name
+}
+
+func (e entity) KeyForType() string {
+	return e.Type
+}
+
+func (e entity) KeyForDescription() string {
+	return fmt.Sprintf("%v", e.Metadata)
+}
+
+func (e entity) Equals(i interface{}) bool {
+	entry, ok := i.(entity)
+	if !ok {
+		return false
+	}
+	return e.Name == entry.Name &&
+		reflect.DeepEqual(e.Metadata, entry.Metadata) &&
+		reflect.DeepEqual(e.Policies, entry.Policies)
+}
+
+func (e entity) CreateOrUpdate(action string) {
+	path := filepath.Join("identity", e.Type, "name", e.Name)
+	config := map[string]interface{}{
+		"policies": e.Policies,
+		"metadata": e.Metadata,
+	}
+	vault.WriteSecret(path, config)
```

Comment:
Assuming there is no tune method available for this mount type?

Reply:
Correct. Both create and update use the same method and parameters in vault api, in which the secret is simply written to the desired path.